datacapturing/src/main/java/de/cyface/datacapturing/backend/CapturingProcess.java

```diff
@@ -284,6 +279,64 @@ public synchronized void onSensorChanged(final @NonNull SensorEvent event) {
+     *
+     * @param eventTimeNanos the timestamp of the {@code SensorEvent} used to determine the offset
+     * @return the offset in milliseconds
+     */
+    long eventTimeOffset(final long eventTimeNanos) {
+
+        // Capture timestamps of event reporting time
+        final long elapsedRealTimeMillis = SystemClock.elapsedRealtime();
+        final long upTimeMillis = SystemClock.uptimeMillis();
+        final long currentTimeMillis = System.currentTimeMillis();
+
+        // Check which timestamp the event.time is closest to, because some manufacturers and models use the currentTime
+        // instead of the elapsedRealTime as the official documentation states for the `event.time`.
+        // Once example could be the Nexus 4 according to https://stackoverflow.com/a/9333605/5815054,
+        // but we also noticed this on some of our devices which we tested back in 2016/2017.
+        final long eventTimeMillis = eventTimeNanos / 1_000_000L;
+        final long elapsedTimeDiff = elapsedRealTimeMillis - eventTimeMillis;
+        final long upTimeDiff = upTimeMillis - eventTimeMillis;
+        final long currentTimeDiff = currentTimeMillis - eventTimeMillis;
+        if (Math.min(Math.abs(currentTimeDiff), Math.abs(elapsedTimeDiff)) >= 1_000) {
+            // Not really a problem but could indicate there is another case we did not think of
+            Log.w(TAG,
+                    "sensorTimeOffset: event delay seems to be relatively high: " + currentTimeDiff + " ms");
+        }
+
+        // Default case (elapsedRealTime, following the documentation)
+        if (Math.abs(elapsedTimeDiff) <= Math.min(Math.abs(upTimeDiff), Math.abs(currentTimeDiff))) {
+            final long bootTimeMillis = currentTimeMillis - elapsedRealTimeMillis;
+            Log.d(TAG, "sensorTimeOffset: event.time seems to be elapsedTime, setting offset "
+                    + "to bootTime: " + bootTimeMillis + " ms");
+            return bootTimeMillis;
+        }
+        // Other seen case (currentTime, e.g. Nexus 4)
+        if (Math.abs(currentTimeDiff) <= Math.abs(upTimeDiff)) {
+            Log.d(TAG, "sensorTimeOffset: event.time seems to be currentTime, setting offset "
+                    + "to zero ms");
+            return 0;
+        }
+        // Possible case, but unknown if actually used by manufacturers (upTime)
+        // If we calculate a static offset between currentTime and upTime this would lead to time
+        // shifts when the device is in (deep) sleep again. We would need to calculate the offset
+        // dynamically for each event which is quite heavy. Thus, we throw an exception to see
+        // in the Play Store if this actually happens on devices. If so, this could be tested using:
+        // https://developer.android.com/training/monitoring-device-state/doze-standby#testing_doze_and_app_standby
+        throw new IllegalStateException("The event.time seems to be upTime. In this case we cannot"
```

Comment:
Is this wise? We might get a plethora of crashes from the field, if we introduce this here.

Reply:
I'm not sure what's better:
1. only release this change to small test groups at the beginning to see if this exception actually exists
2. handle this softly and never know if this actually happens. Soft handling would mean (A) static offset --> data might be shifted multiple times or (B) re-calculate offset on each sensor update (or a selection of those), without knowing if this is necessary or (C) ...

I would go (and went for): 1.
As long as we introduce this slowly (own test phones, research partner phones, public test app users) we actually get a good sense if this is a exception which can occur or not and can eventually release this change to paid customer's users.